### PR TITLE
dec_sub: cache up to 10 packets

### PR DESCRIPTION
**Pull request by Dudemanguy**
062104d16e34f348ffd9324ca4c997b6b0f487d4 started saving 2 packets in memory for the purposes of runtime updates of some options, but sometimes we need more. There are subs out there that will have overlapping durations so saving only 2 packets would only update those 2 subs at most and drop the rest. So with that in mind, let's just cache up to 10 (arbitrary but surely no one needs more) packets instead. We don't always want to hold onto that many since it's useless 99% of the time, so do some logic by checking the video pts and the durations. At least 2 must always be kept though so there's some extra conditions.

### CODE REVIEW

File: sub/dec_sub.c [hunk 1]
```diff
@@ -193,6 +195,32 @@ struct dec_sub *sub_create(struct mpv_global *global, struct track *track,
...
+    // Always save at least 2 packets but allow up to PKT_CACHE
+    // for subs with overlapping durations.
+    int i = PKT_CACHE;
+    bool extra_pkts = false;
+    while (--i) {
+        struct demux_packet *prev_pkt = sub->cached_pkts[i - 1];
+        if (prev_pkt) {
+            extra_pkts = i > 1;
+            double prev_pkt_end = prev_pkt->pts + prev_pkt->duration;
+            if (sub->cached_pkts[i] && prev_pkt_end < video_pts && (i > 1 || !extra_pkts))
```
Comment (christoph-heinrich) [inline sub/dec_sub.c:215]: `(i > 1 || !(i > 1))`

File: sub/dec_sub.c [hunk 2]
```diff
@@ -193,6 +195,30 @@ struct dec_sub *sub_create(struct mpv_global *global, struct track *track,
...
+    // for subs with overlapping durations.
+    int i = PKT_CACHE;
+    while (--i) {
+        struct demux_packet *prev_pkt = sub->cached_pkts[i - 1];
+        if (prev_pkt) {
+            double prev_pkt_end = prev_pkt->pts + prev_pkt->duration;
+            if (sub->cached_pkts[i] && prev_pkt_end < video_pts)
+                TA_FREEP(&sub->cached_pkts[i]);
+            sub->cached_pkts[i] = prev_pkt;
+            sub->cached_pkts[i - 1] = NULL;
```
Comment (christoph-heinrich) [inline sub/dec_sub.c:216]: I'm probably not understanding it correctly, but shouldn't that go in an else branch?

If `sub->cached_pkts[i] && prev_pkt_end >= video_pts` then `sub->cached_pkts[i]` doesn't get freed but it still gets overwritten by `sub->cached_pkts[i - 1]`.

Comment (christoph-heinrich) [reply sub/dec_sub.c:216]: But then it's not guaranteed that `sub->cached_pkts[0]` is free for another packet...

Comment (Dudemanguy) [reply sub/dec_sub.c:216]: >it still gets overwritten by sub->cached_pkts[i - 1].

That's was intended to always advance the cache but indeed it never frees anything until the cache maxes out.

**Comment by mitzsch** [comment]
Sorry for the (dumb) question. Does this also affect picture-based subs (like pgs) or only text-based subs (ass, srt etc)?

**Comment by Lypheo** [comment]
Only text in practice (both are cached, but with picture subs the cache is never used).

**Comment by Dudemanguy** [comment]
#13339 incidentally handles this is in a far better way and solves this particular problem. So closing.

**Pull request closed without merging**